10_rest/willshen8/cmd/puppy-server/main.go

```
@@ -0,0 +1,118 @@
+	storer   = kingpin.Flag("store", "Map/Sync Store").Short('s').Default("map").String()
+)
+
+// Config contains arguments parsed from commandline flags.
+type Config struct {
+	file   string
+	port   int
+	storer string
+}
+
+func main() {
+	cmdConfigs := parseCmdArgs(args)
+	newStorer, _ := createStorer(cmdConfigs)
+	decodedPuppies, _ := parseFileFlag(cmdConfigs)
+	portFlag, portErr := parsePortFlag(cmdConfigs)
+	if portErr != nil {
+		fmt.Println(portErr)
+	}
+	port := ":" + portFlag
+
+	puppyStorer, _ := createPuppies(newStorer, decodedPuppies)
+	fmt.Fprintln(out, puppyStorer)
+	puppyHandler := puppy.NewRestHandler(puppyStorer)
+	router := SetupRouter()
+	puppy.SetupRoutes(router, *puppyHandler)
+	log.Fatal(http.ListenAndServe(port, router))
+}
+
+// creatStorer will take the commandline arg and create the appropriate store
+func createStorer(c *Config) (puppy.Storer, error) {
+	switch c.storer {
+	case "map":
+		return puppy.NewMapStore(), nil
+	case "sync":
+		return puppy.NewSyncStore(), nil
+	}
+	return nil, errors.New("map/sync are the only acceptable flag values")
+}
+
+func parseCmdArgs(args []string) *Config {
+	if _, parseError := kingpin.CommandLine.Parse(args); parseError != nil {
+		panic(parseError)
+	}
+	return &Config{file: *fileName, port: *port, storer: *storer}
+}
+
+func parseFileFlag(c *Config) ([]puppy.Puppy, error) {
+	var file *string
+	var puppies []puppy.Puppy
+	file = &c.file
+	jsonData := readFile(*file)
+
+	if err := json.Unmarshal(jsonData, &puppies); err != nil {
+		return nil, err
+	}
+	return puppies, nil
+}
+
+// parsePortFlag check for valid port number entered from commandLine
+func parsePortFlag(c *Config) (string, error) {
+	if c.port < 0 || c.port > 65535 {
+		return "8888", errors.New("invalid port number entered, default port 8888 will be used")
+	}
+	return strconv.Itoa(c.port), nil
+}
+
+func readFile(file string) []byte {
+	jsonData, err := ioutil.ReadFile(file)
+	if err != nil {
+		panic(err)
```

why not return an error here?
I'd foldit into parseFileFlag.

---

Outdated, this funtion no longer exists.